virttest/test_setup.py

```diff
@@ -97,6 +100,90 @@ class PolkitConfigCleanupError(PolkitConfigError):
     pass
 
 
+class Setuper(object):
+
+    """
+    Virtual base abstraction of setuper.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self.test = test
+        self.params = params
+        self.env = env
+
+    @abstractmethod
+    def setup(self):
+        """Setup procedure."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def cleanup(self):
+        """Cleanup procedure."""
+        raise NotImplementedError
+
+
+class SetuperMgr(object):
+
+    """
+    Setuper Manager implementation.
+
+    The instance can help do the setup stuff before test started and
+    do the cleanup stuff after test finished. This setup-cleanup
+    combined stuff will be performed in LIFO order.
+    """
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper manager.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self._stp_args = (test, params, env)
+        self._reg_stps = []
+        self._setupers = []
+
+    def register(self, stp_cls):
+        """
+        Register the given setuper class to the manager.
+
+        :param stp_cls: Setuper class.
+        """
+        self._reg_stps.append(stp_cls)
+
+    def do_setup(self):
+        """Do setup stuff."""
+        for stp_cls in self._reg_stps:
+            stp = stp_cls(*self._stp_args)
+            stp.setup()
+            self._setupers.append(stp)
```

Comment:
Even failed `setup` means some steps might have been executed, so the cleanup should be executed as well. So the `append` should happen before the `setup()` call to allow partial set cleanups.

Reply:
I agreed that the "cleanup" should be executed even if setup failed, but it must be done within `setup` because `setup` is executed in `preprocess` and `cleanup` is executed in `postprocess`, we can't let tests run on such a dirty environment.

Actually, the whole steps in `setup` should be performed atomically, either all pass or all fail.

Reply:
Well it's always a tough call. Right now it works the way that setup usually fails immediately but cleanup should not interrupt. I'd keep it that way first as some setup steps depend on previous ones (like cloning images)